### PR TITLE
Fixes #1004: Support Swagger tags

### LINKED ISSUE #1004: [Feature Request] Add group endpoints by tags for swagger
- Insomnia Version: 5.16.6
- Operating System: Win10 x64

## Details
Hi, 

I like to use Insomnia because UI is simple effective ! but an option is missing: "I would like to see requests grouped by tags"
 
## Example

1. Go to: [http://petstore.swagger.io/#/pet](http://petstore.swagger.io/#/pet ). You can see 3 endpoints ( pet; store and user). 
2. Import json API definition in Insomnia (Import/Export Data --> Import Data --> From URL and set : https://petstore.swagger.io/v2/swagger.json)
3. You can see on left panel requests aren't grouped :(    it's unreadable
image

## Goal
I think, i will be better if requests are grouped by tags (pet; store; user) , something like that :  (
image

### PR DESCRIPTION
**Pull request by thewheat**
Closes https://github.com/getinsomnia/insomnia/issues/1004

- Adds support for Swagger tags which groups requests into folders

image

- Also provides support for multiple tags in a single request: it will create duplicate requests to list them in multiple folders
   - A request with multiple Swagger Tags (i.e. Insomnia folders) is possible and displaying multiple times is expected as seen [in this Swagger UI issue](https://github.com/swagger-api/swagger-ui/issues/3711)
   - Sample below is a [modified json file](https://github.com/getinsomnia/insomnia/files/2475114/swagger_other_cases.json.txt) where 
      - `Update an existing pet` has tags of `pet` and `store` and 
      - `Add a new pet to the store` has no tags

image

### CODE REVIEW

File: packages/insomnia-importers/src/importers/swagger2.js [hunk 1]
```diff
@@ -100,28 +104,73 @@ function parseEndpoints(document) {
...
+  const folderLookup = {};
+  folders.forEach(folder => (folderLookup[folder.name] = folder._id));
+
+  const requests = [];
+  endpointsSchemas.map(endpointSchema => {
+    let { tags } = endpointSchema;
+    if (!tags || tags.length == 0) tags = [''];
+    tags.forEach((tag, index) => {
+      let id =
+        endpointSchema.operationId + (index > 0 ? index : '') || `__REQUEST_${requestCount++}__`;
```
Comment (thewheat) [inline packages/insomnia-importers/src/importers/swagger2.js]: Added a suffix with index as if the request has multiple tags, each request needs a unique id and thus append index to make it unique

Comment (gschier) [reply packages/insomnia-importers/src/importers/swagger2.js]: Nice, that sounds like a good solution 👍

**Review by gschier — APPROVED** [review]
This looks awesome! This has been requested for a long time so thank you for taking care of it 🎆

**Comment by mawiseman** [comment]
What happened to this pr? seems to have never made it in?

**Comment by thewheat** [comment]
Hi @mawiseman 👋 I've imported https://petstore.swagger.io/v2/swagger.json and correctly puts the requests in different folders (each folder represents the tag on the request) so it looks like it has been merged.

Do you have a URL / file where it doesn't work? That could help us figure out what's wrong 👍 

image


These are my Insomnia details if it helps as well 😃
```
Version: Insomnia Core 2020.4.1
Release date: 9/23/2020
OS: Darwin x64 19.6.0
Electron: 9.1.1
Node: 12.14.1
V8: 8.3.110.13-electron.0
Architecture: x64
node-libcurl: libcurl/7.69.1 OpenSSL/1.1.1g zlib/1.2.11 brotli/1.0.7 libidn2/2.1.1 libssh2/1.9.0 nghttp2/1.41.0
```

**Comment by mawiseman** [comment]
Thanks for looking into this @thewheat 

I have attached a cut down version of the swagger file I am trying to import.
[swagger.json.txt](https://github.com/Kong/insomnia/files/5369233/swagger.json.txt)


It was generated using the Swashbuckle library for dotnet core: https://github.com/domaindrivendev/Swashbuckle.AspNetCore

You can see it is valid when pasting it into the online editor: https://editor.swagger.io/


Insomnia Import & Swagger View

image

image

Here are my Insomnia details

```
Version: Insomnia Core 2020.4.1
Release date: 9/23/2020
OS: Windows_NT x64 10.0.18363
Electron: 9.1.1
Node: 12.14.1
V8: 8.3.110.13-electron.0
Architecture: x64
node-libcurl: libcurl/7.69.1-DEV OpenSSL/1.1.1d zlib/1.2.11 WinIDN libssh2/1.9.0_DEV nghttp2/1.40.0
```

**Comment by thewheat** [comment]
Ah @mawiseman I see that your file says it is an `OpenAPI` specification and not a `Swagger` one (Insomnia currently has a different import for each of those and the one I modified was Swagger 😅
image

**Swagger definition [in the Petstore example](https://petstore.swagger.io/v2/swagger.json)**
image

Did a test and if you add tags in the root definition it does correctly create the folders
image

Digging further, from https://github.com/Kong/insomnia/issues/2408 it looks like the beahviour of needing to add it to global scope is expected. Not ideal but if anything, I think that issue would be best for any further discussion 👍

**Comment by mawiseman** [comment]
Thanks @thewheat!

**Pull request merged**